### ISSUE TITLE
Duplicating a node should result in either a true Instance or true unique copy, not a mix of both

### ISSUE THREAD

**Issue by golddotasksquestions** [godotengine/godot-proposals#317]
**Describe the project you are working on:**
In all my projects I'm duplicating and instancing nodes all the time.

**Describe the problem or limitation you are having in your project:**
When I duplicate nodes, they are neither unique (aka independent or disconnected from their original), nor are they completely connected: a pure instance where changing one thing in any one of the duplicates results in the same change happening to all of the duplicates.
![instance_or_duplicate](https://user-images.githubusercontent.com/47016402/71084030-a8daa800-2194-11ea-97b2-2bd72f6e2fb6.gif)

**Describe how this feature / enhancement will help you overcome this problem or limitation:**
Having either true Instances of nodes or true unique duplicates would prevent confusion, unintentional changes, self inflicted bugs.
Ideally I would like to be able to do something like
rightclick on a node in the Scene Panel for the context menu and select "create Duplicate (Ctrl+D)" to create a unique duplicate
and "create Instance (Ctrl + I )" (or Ctrl+Shift+D) to create a true Instance of the Node I just created.

If I would want a specific Resource to be shared, by multiple nodes, I would like to Save the resource as a .res for instance, and then load it into any another node (already possible).
This would be intentional driven design for me.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
Imagine either both color changing _and_ shapes changing in the gif above (true instance) or both color _and_ shape changes stay independent from the original (unique duplicate)

**Describe implementation detail for your proposal (in code), if possible:**
for me not possible

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
Would be used a lot because duplication of nodes is a very common thing in Godot.

**Is there a reason why this should be core and not an add-on in the asset library?:**
It's about existing core functionality

**Comment by Xrayez** [comment]
Setting up resources to be unique across instances is possible by enabling `Local To Scene` option:

![godot-local-to-scene-duplicate](https://user-images.githubusercontent.com/17108460/71091901-bb61db00-21ae-11ea-8a59-a8c081146e89.png)

It works for scenes only though. I wish I knew about this feature from the moment I started using Godot!

Resources are shared by default to improve re-usability and performance. Scripts are also resources and they 99% of the time have to be shared. One can take a conscious advantage of this so that changes to a single resource can affect some behavior of existing nodes at run-time so you don't have to go through each node to change a particular property (say friction, see [PhysicsMaterial](https://docs.godotengine.org/en/latest/classes/class_physicsmaterial.html)).

I think there's much confusion as to how duplicating works in Godot, for instance see godotengine/godot#33824.

So yeah I think it's necessary to distinguish between shallow/deep/instance duplication across the editor. Adding an additional submenu popup for this could be a step forward:

![godot-duplicate-unique](https://user-images.githubusercontent.com/17108460/71092908-cc135080-21b0-11ea-9c54-9753a8355e27.png)

**Comment by Zireael07** [comment]
Related to @Xrayez's comment: https://github.com/godotengine/godot/issues/6258, https://github.com/godotengine/godot/issues/6922, https://github.com/godotengine/godot/issues/9037

**Comment by golddotasksquestions** [comment]
Thank you @Xrayez  for sharing your thoughts!

I totally get why it is good and important to allow resources to be shared.
What I don't get is the fact that this is the default. It currently is opt-out, but strongly feel it should be opt in.

Or clearly differentiate between instancing (everything shared) and duplication (everything unique).

The shapes resource is just a visual example you can easily miss until many many commits later. I actually have and had much more problems with scripts and animations before I learned never to share scripts and never to save animations. 

It should not be like that. Sharing scripts with nodes should not threat to override your work if you did not know of forgot to tick that checkbox.

**Comment by Feniks-Gaming** [comment]
I agree with the naming change. Duplicate doesn't suggest the linked copy. When I duplicate the file on computer it doesn't change because I change original or more importantly even original doesn't change because I edited a copy. Separating instance and duplicate would be clearer

**Comment by willnationsdev** [comment]
@Calinou I believe the `topic:core` label is also relevant here as they are suggesting that these changes also be applied even when executed from code. It isn't an editor-specific issue.

**Comment by jasonwinterpixel** [comment]
If we implement this proposal, would that help with your issue?

https://github.com/godotengine/godot-proposals/issues/2737

The workflow would be Duplicate Node -> Make Subresources Unique

**Comment by chottokite** [comment]
Duplicating a node merely duplicates its node tree. Perhaps it doesn't do what you expect it to with some idea on mind but, it does what it says it does and not more. This behavior should stay as is. When in the editor, I think you should consider making these nodes you duplicate a separate scene and instance them in the main scene. If needed, mark those particular resources that you want to make unique with "Make local to scene". If duplicating nodes in GDScript, you should consider again separate scenes instances and duplicating nodes with the DUPLICATE_USE_INSTANCING flag.

**Comment by jasonwinterpixel** [comment]
> This behavior should stay as is. 

My proposal https://github.com/godotengine/godot-proposals/issues/2737 keeps duplicating a node the same. I'm referring to the behaviour of Make Subresources Unique which only operates on 1 layer of subresources. I'm suggesting that Make Subresources Unique should truly make all the subresources unique.

**Comment by jordanlis** [comment]
> 
> 
> Duplicating a node merely duplicates its node tree. Perhaps it doesn't do what you expect it to with some idea on mind but, it does what it says it does and not more. This behavior should stay as is. When in the editor, I think you should consider making these nodes you duplicate a separate scene and instance them in the main scene. If needed, mark those particular resources that you want to make unique with "Make local to scene". If duplicating nodes in GDScript, you should consider again separate scenes instances and duplicating nodes with the DUPLICATE_USE_INSTANCING flag.

The important thing to understand is not that the software should do what it says it will do. You have to be in empathy with the user. Here is at least one use case that leads to say that ctrl + D should create unique instances : 
- When you instanciate a new scene, let's say a tree in your level, you start by doing it
- Then, you probably have multiple trees in your levels, so you don't wan't to go through the whole process to instanciate your next trees, because it takes time and that you want to do your level design fast. So you select the node and make control + D

In this process, you will have exactly duplicated nodes, but you want to have true differents instances because each tree can be configurable with for examples values in the editor.
SO : if you used ctrl + D, and that you modify a parameter of one tree in the inspector, it will affect all the others trees whereas it is not what is expected by the level designer, because he want's to create chaos like in nature and but different parameters for each tree.

The user experience is the most important, not the pure logic of the software because it's used by human beings

**Comment by lukostello** [comment]
I find this post confusing because there are 4 ways it could be implemented and I'm having track of which methods are being called what.
1. When you make a copy it wouldn't share resources if "make_local" on that resource is checked (then either A: recursive subresources arn't shared as well if they are checked as "make_local" or B: Subresources even when marked "make local" will still be shared among all the unique parent resources that were made) but do share resources otherwise. This (B version at least) is what happens when you instance a packed scene right now so I hope this is what is being called instance. But sometimes it seems like the next one is what other people are talking about when they say instance
2. When you make a copy it not only would it share resources (even ones that weren't made local) but every variable the would be the same as well. As the OP seems to expect seeing as how canvas item is inherited from the base class not a resource. But this simply wouldn't make sense because if we did that they would have to share translation data as well which would make them look like one object. 
3. Similar to 1 but as though all resources (and subresources) are not checked make local (even if they are)
4. Similar to 1 but as though all resources (and subresources) are checked make local (even if they aren't)

I'd hope 1. would be called instancing because its what happens when we instance a packed scene. I'd expect duplicate to mean 4.

**Comment by ExquisiterEmil** [comment]
As a level designer, I cannot describe how often the current duplication behavior has been driving me crazy, especially since I do not know such behavior from any other 3D software. To give you a perspective on the importance of fixing this here is an example of how annoying this is:
I have an area that defines camera movement. It contains 3 Collision shapes for: 
- confining the camera
- detecting when the player has entered the area
- detecting when the player has exited (usually bigger than the "entered" area, to avoid switching too fast between scenes, or continuous switching when the player is on the edge)
Now when want to add such a zone, I often duplicate an existing one, since it is probably not far away in the viewport and has similar sizes. I then have to make three resources unique every time I make a new area. This is time-consuming. But the worst part is, that it is very easy to forget, and when the old area is not in view any more or hidden, there is no way of noticing, that you are overriding carefully placed colliders. Repeat this a few times and you can lose several minutes of work and feel very frustrated. This problem applies to a multitude of similar systems. I think for beginners, this behavior is particularly unintuitive and off-putting. 

One way of fixing this could also be to have a setting on a resource field, next to "Local to Scene" that would say "Make unique on duplication". Reading the solution from @lukostello, I want to mention that this might not only apply to objects that have been "Made local", but also for objects with "Editable children" applied.

**Comment by beyarkay** [comment]
What is the current state of this proposal? Has it been accepted, or does it need more work to clarify what exactly will change and how that change will be implemented? I'm also being bitten by the fact that doing a deep copy isn't easily available.

**Comment by Calinou** [comment]
> What is the current state of this proposal? Has it been accepted, or does it need more work to clarify what exactly will change and how that change will be implemented? I'm also being bitten by the fact that doing a deep copy isn't easily available.

We don't know what a definitive API would look like, considering this should be available both from code and from the editor. It's also not certain whether the default behavior for <kbd>Ctrl + D</kbd> in the editor should be changed. That said, I would prefer not changing the default behavior and adding an alternative shortcut instead, as duplicating resources without actually needing to can lead to a performance (and memory usage) penalty.

**Comment by jordanlis** [comment]
Controversial maybe, and much more about the solution to apply. But in the issue you linked, many people appears to agree on the fact that things are really confusing at the moment about shared ressources and the default behavior of copy paste of a node.
Maybe small things can be done to fix it, but I think a lot of people at least agree that a change somewhere should be done no?

**Comment by luislodosm** [comment]
For me, the expected behavior when using `ctrl+d` or `copy/paste` on a node is:
- If the node has references to `external resources` stored on the Filesystem, I expect the references to be maintained.
- If the node has `built-in resources`, I expect new internal resources to be created in the new node.

Fortunately, Godot is very open, so my solution is this:

```
@tool # <- Important
class_name MyClass

@export var my_external_resource: MyResource
@export var my_built_in_resource: MyResource

func _enter_tree():
	var im_duplicating = Input.is_action_just_pressed("ui_graph_duplicate")
	var im_pasting = Input.is_action_just_pressed("ui_paste")
	if im_duplicating or im_pasting:
		my_built_in_resource = MyResource.new()
```

**Comment by Malleator** [comment]
I still can't believe how many years that this issue has sat here; furthermore, I can't believe that, 'Duplicate,' makes a reference instead of a unique actual duplicate in the first place. Seems like an incredible oversight in design. 99% of the time the user is gonna want and expect a duplicate, not a reference. And the fact that the only alternative, seemingly at this time, to actually duplicating a node is manually recreating it is simply preposterous in my opinion

**Comment by golddotasksquestions** [comment]
@Malleator 

What is confusing, is that duplicate does make an actual duplicate, but only of the node itself, not of the Resources this node uses.

So to not be completely confused and fucked by this issue, you first have to have a really good understanding of what Godot means by "Resource" and how this is different from the node.

_Nodes use Resources._

Take a Sprite for example. When we see a graphic in a 2D game most people would call this a Sprite. When in fact what we see is a texture (= Resource) displayed by a Sprite Node. If the user duplicates this Sprite, Godot will create a new Sprite instance from the Sprite class and assign it the same Resource by reference. Meaning the user can change all Sprite properties individually for each Sprite, but as soon as the user fiddles with the Resource of one Sprite, all Sprites will suddenly adopt these changes.

Now for a Sprite using a Texture resource this might be desired behaviour. For a CollisionShape using a Shape resource however, or an AnimationPlayer using an Animation resource, or a Node using a Script resource, ... this might not be desired in the most cases.

The biggest problem Godot has in this regard imho, is the silence. The total lack of transparency and feedback in communicating this to the user via the Editor GUI. You basically have to be an expert Godot user to fully understand this behavior which is very specific to Godot. Would it be more transparent (for example with **a giant icon and large fat text** next to the resource in the Inspector **making it crystal clear this resource is shared or unique**, I think this issue would be much less of a problem.

**Comment by jordanlis** [comment]
> @Malleator
> 
> What is confusing, is that duplicate does make an actual duplicate, but only of the node itself, not of the Resources this node uses.
> 
> So to not be completely confused and fucked by this issue, you first have to have a really good understanding of what Godot means by "Resource" and how this is different from the node.
> 
> _Nodes use Resources._
> 
> Take a Sprite for example. When we see a graphic in a 2D game most people would call this a Sprite. When in fact what we see is a texture (= Resource) displayed by a Sprite Node. If the user duplicates this Sprite, Godot will create a new Sprite instance from the Sprite class and assign it the same Resource by reference. Meaning the user can change all Sprite properties individually for each Sprite, but as soon as the user fiddles with the Resource of one Sprite, all Sprites will suddenly adopt these changes.
> 
> Now for a Sprite using a Texture resource this might be desired behaviour. For a CollisionShape using a Shape resource however, or an AnimationPlayer using an Animation resource, or a Node using a Script resource, ... this might not be desired in the most cases.
> 
> The biggest problem Godot has in this regard imho, is the silence. The total lack of transparency and feedback in communicating this to the user via the Editor GUI. You basically have to be an expert Godot user to fully understand this behavior which is very specific to Godot. Would it be more transparent (for example with **a giant icon and large fat text** next to the resource in the Inspector **making it crystal clear this resource is shared or unique**, I think this issue would be much less of a problem.

I understand your point of view, but imho your overcomplicate things. There are so many simple solutions and different level of answer for this user's needs. 

Level 0 : Explain to the customer in some way how it works when duplicating a node (resource will not be duplicated, but reference will be kept for each node). 
Level 1 : **Ctrl + D** is duplicating the node, **ctrl + MAJ + D** will duplicate the node AND its ressources. And we can add "Duplicate node and ressources" entry in the sub-menu.
Level 2 : Ability to set default behavior for this feature in the project settings. See this proposal https://github.com/godotengine/godot-proposals/issues/3072
Level 3 : managing resources for all the nodes of a scene in a pop-in
Level 4 : Add wathever you can imagine ^^ 

In fact, the issue here is not wether this user's need is relevant or not. Many people are convinced that this IS relevant (but not very critical maybe). It is only a question of wether a developer would like to make any of these modifications above or not, because this project is open source so I guess we just have to wait for a developer to make a small modification to answer this issue.

**Comment by maridany1999** [comment]
I came across the same issue while duplicating nodes via code. Duplicating a parent node, especially one with exported variables, retains references from the original parent node, which I consider a bug.

I've observed similar problems in other related issues here in GitHub, where users are confused, negatively impacting the user experience (UX) for some. Currently, I can only see a few practical uses using this behavior. In most software applications, duplicating something typically creates a unique copy.

A potential solution could involve offering two options: "Duplicate as New" and "Duplicate (Use Original References)". This approach would not only help users understand the intended functionality, but also improve the overall UX (similar to the "Duplicate" and "Duplicate Link" function in Blender, but a little different in this case from what I understand).

**Comment by CoyanCardenas** [comment]
As someone who is new Godot, I just want to say I've lost many hours of work due to this. It is counter-intuitive to what I believe most new users will expect when they duplicate something. Although there are times when you want shared data between the two - you would never want a script duplicated on two nodes for example - I think the majority of the time the expected behaviour is that a duplication is completely seperate and changes to one wont affect the other.
If I want to have instances, where changes made to the base will affect all instances, I will create a new scene and *instantiate* that scene. Knowing full well that any change to the base will change all, and any change to individual instances only affects that specific instance.
The fact that the current state of things means people can easily and unknowingly lose work, even if they understand how it works but just forget to check local to scene on every single sub resource, is testiment to something needing to be done about this.

**Comment by Kalto-Mate** [comment]
I think the biggest offender here is the user being betrayed by terminology. In any other situation when working with a computer, duplicating something with Ctrl + D(uplicate) creates an independent copy of whatever was duplicated. 

If I duplicate a document that references external images, I expect edition of font size in the copy to just happen there while those external images remain existing only once. If I duplicate a node tree I expect the resources it may reference to be shared, but not internal settings like bounding size to be entangled. Specially if they are like this silently with no indication about them being connected.

Regardless of the debate about what should be the default, a non-independent copy umbilically connected to it's brothers should absolutely be called an INSTANCE, not a duplicate, and the context menus and even keyboard shortcuts should reflect this nuance. I too have lost work to this intuitiveness and sneaky undisclosed behavior behind my back.

At the very least add a way to quickly make true unconnected duplicates of stuff If we are this dead set on having the main default way be secretly instances but enjoy the term "duplicate"

**Comment by JackVania** [comment]
I agree with what everyone else is saying. "Duplicate" does not imply "Instance", "Shared" - it implies a unique copy - because that's literally how everything else works. If we create copies of our applications in various states as forms of backup, we do not expect any changes we make to the current to be applied to the former backup. That is absurd. This isn't even about game engines, this is about consistency of expectation that is shared across multiple pieces of software going back as far as I can remember.

In a DAW if I copy a piano roll and make edits to a single measure, I do not expect those changes to apply to previous piano rolls in earlier measures. No software does this. When you copy or duplicate it becomes its own unique entity. Imagine if Photoshop does this with layers, DAWs do this with edits, NLE's do this with video editing, effects, timeline - mass hysteria.

A LOT of what Godot does changes the game for development in a good way but there's corners here and there that make absolutely no sense in any context you try to give it. An actual "duplicate" as everyone else understands because they've been making duplicates or copy/pasting across multiple pieces of software for _decades_ is the expectation, the current behavior is not and does not improve the workflow in any way, it hinders it.

**Comment by Cyan232** [comment]
same here, would it be possible to add a enum DuplicateFlags, to chose if a copy or copy of instance is duplicated ?
https://docs.godotengine.org/en/stable/classes/class_node.html#enum-node-duplicateflags

**Comment by pc9098** [comment]
![image](https://github.com/godotengine/godot-proposals/assets/78771410/6a1bcd64-329f-4b18-a7ae-4205827bb309)
I'll propose something like this

**Comment by PLyczkowski** [comment]
Maybe a popup when duplicating?
![image](https://github.com/godotengine/godot-proposals/assets/7307968/3bd4d7eb-1575-43a0-8db3-4da055e54fe1)

**Comment by Calinou** [comment]
> Maybe a popup when duplicating?

Duplicating is a frequent operation, so I don't think it should spawn a dialog every time you try to duplicate something.

At most, it should display a toast to mention a resource is shared when you duplicate it (specifically if we go for https://github.com/godotengine/godot-proposals/issues/4672).

**Comment by amarraff** [comment]
I ran into this problem today when working with UI. Users in the Godot Discord were also confused about how to fix it when I asked for help. 

I have a "button" scene with a script attached that assigns information to button labels based on other info in my project. I dragged the button scene into the node hierarchy, and then duplicated it twice. So, there was "button," "button2," and "button3." On _ready(), my code for all three buttons ran, but applied everything to the original/first button's labels only. This default behavior is extremely confusing. 

Echoing what others here have said, duplication or copy/pasting should not create a linked instance; it should create a unique copy, similar to just dragging the scene into the hierarchy multiple times (which fixed the issue for me). Providing a separate button/shortcut for "Make Linked Instance" would alleviate a lot of confusion.

**Comment by pc9098** [comment]
Since [making a new button to create an independent duplicate](https://github.com/godotengine/godot-proposals/issues/317#issuecomment-1874569807) isn't a popular option
and [popping a window where you can select which resources should be unique](https://github.com/godotengine/godot-proposals/issues/317#issuecomment-1877487060) isn't a popular option
I propose making this option in the editor's option like this:
![image](https://github.com/godotengine/godot-proposals/assets/78771410/0be84cdb-4b9a-4a23-b4d1-b0fa96fb35c4)

**Comment by amarraff** [comment]
Just ran into another issue that appears to be linked to this issue, with resources. I have two resource types - for simplicity, let's call them apple and pear.

I had an apple in my file system, and I wanted to create a pear. So, I duplicated my apple resource, removed the "apple.gd" script, and added pear.gd. So now, functionally, from the end-user's perspective, it has been converted into a pear.

Unfortunately, when I went to "quick load" the pear resource for an export in the inspector, it wouldn't show up. Only once I deleted the pear, created a new pear resource from scratch, and assigned pear.gd to it, did it finally show up in quick load.

**Comment by desastreger** [comment]
Hi all - Someone posted something about this on Reddit and I think the solution could be fairly simple:

- In the editor, In the context menu whilst clicking on a node, replace Duplicate with "Duplicate and link" and "Duplicate and make unique"
- Create a friendly yield icon for linked nodes indicating that it is not unique, once hovered on it can display a tooltip with all the linked nodes

I feel that if we havea default setting we're always going to be thinking about the "toggled off" option when working in the editor.

That's it, fairly clean.  Thoughts?

**Comment by danchaseCTS** [comment]
What is the workaround while this is being argued? Do you have to just add each resource and re-do it?

**Comment by amarraff** [comment]
> What is the workaround while this is being argued? Do you have to just add each resource and re-do it? In my case my CollisionShape2D's all resize when I try to resize 1.. I just used duplicate as a shortcut, but do I just need to delete them all and re-add them each as new?

To my knowledge, yes. In some instances duplicating will just create unique instances, which makes this even more confusing. 

In case you weren't aware, you can right click on components of a node and hit "make unique."

For example, this will work with the mesh of a MeshInstance3D; make one, assign its shape, and duplicate the node. They'll be linked. Right click the second one's mesh and hit make unique. They'll no longer be linked.

I'm not sure if there's an indication anywhere that they're linked. I'd like to be able to view that information, if it's there.

**Comment by danchaseCTS** [comment]
@amarraff Thank you for clarifying!

**Comment by daniel080400** [comment]
I propose a breaking change, in my view this might work:

Change the resource bool property "local to scene" to "prefer shared copy", to have the scope focus on the resource itself. If this is set to true, no matter the resource is internal or external, either duplication or instantiation, this resource will always result in a linked reference. This option will only take effect when it's created, users can still make the resource unique afterwards.

With this "prefer shared copy" option, we can have default values for various resources. For example, most internal resources should have this set to false by default, with few exceptions like built in scripts; most external resources (saved to disk as files) should set to true. Users are free to toggle this option as they wish, but specific resources like textures and meshes will display a obvious warning when user try to toggle "prefer shared copy" to false.

**Comment by danchaseCTS** [comment]
Hey just FYI from my opinion, I was in favor of wanting the shared functionality to take a back seat, until I was in the middle of a game and realized the impact of what I was suggesting. My opinion it's easy to think of this as an editor like MS WORD or Photoshop, Blender, or something where _generic functionalities_ add up to usability, but this is not a word document, in the end things have to work together and if things didn't share resources you'd have to go back later and start linking things you should have linked from the beginning. Tools that don't do linking, just put it on you to go back through everything and make sure it's all right. Was nice that Godot did it for me, and when I realized it, I finally understood more about this whole argument.

**Comment by pc9098** [comment]
Any way this can be prioritized? This is a problem I'd say everyone has run into and it's been present for 5 years

**Comment by Calinou** [comment]
> Any way this can be prioritized? This is a problem I'd say everyone has run into and it's been present for 5 years

https://github.com/godotengine/godot/pull/103980 adds as a **Paste as Unique** option to the resource dropdown menu. This should significantly speed up workflows already.

Further changes are more involved, especially considering we can't break compatibility just yet. [Detecting the number of resources using another resource](https://github.com/godotengine/godot-proposals/issues/904) is complex. I've also tried to implement https://github.com/godotengine/godot-proposals/issues/4672#issuecomment-1792913954 but as you can see, I've run into issues that prevent the feature from working as intended.

**Comment by shoeffner** [comment]
@danchaseCTS For the original question (i.e., Shape resources of CollisionShape2D nodes), I have a small workaround in https://github.com/godotengine/godot/issues/49916#issuecomment-2840183058 . It duplicates a (selected) CollisionShape2D and then its Shape resource onto the new CollisionShape2D when you press Ctrl+E.

**Comment by MajorMcDoom** [comment]
I know users who encounter this issue argue that linking is a bad default behaviour for duplicating, but I disagree. Though I totally agree on the severity of this issue.

The key difference between Godot and most engines which causes frustration in most of the scenarios mentioned is that _so many_ things in Godot are _resources_, where they would be simply "nodes" or "Components" in other engines. That is, Godot's very architecture promotes heavy reuse. This isn't always a good thing.

A prime example of this is a collision shape. In other engines, the "collision shape" is just what results from the properties on a collision component. Godot makes this a reusable resource. I would argue that while this decision has some fringe benefits, it is almost always a bother. Collision shapes are for the most part unique, and do not take up a lot of memory, making them poor candidates for "resourcification". Another example is curves and paths.

On the other hand, something like a material should absolutely be reused by default. Imagine if you made 100 copies of an object that has 2 materials, and Godot silently created 200 unique materials instead of just reusing the same 2 materials. That would be disastrous for performance.

As such, there is actually no fundamental difference in Godot's _duplication logic_ vs. that of any other engine. It has always been "copy value if it's a value", and "copy reference if it's a reference". The difference is that Godot makes most things a reference (resource). 

Unfortunately for people who dislike this behaviour, it really muddies the water, because you can't just draw an arbitrary line around _some_ resources to make a "true duplicate". And forcing a global change in default behaviour might have the superficial effect of making this odd UX problem go away, but it's actually a further deviation from engine norms, because it'll duplicate a lot of resources that you _do_ expect to be linked.

**Comment by CookieBadger** [comment]
@MajorMcDoom I think we are getting to the core of the issue here. I suspect that this pattern (specifically for collision shapes) was likely the result of reusing the Resource functionality for something that it was not made for. I'll explain:

You want to create a new node that has some properties, like the debug color of a collision shape. You can add those by binding them, and they will show up in the inspector. Now you want your node to have different *modes* like the actual shape of the collider (square, triangle, circle, etc.), and depending on that mode, you have properties. Now, when the feature was developed the thought process was likely "I can create an abstract resource, and create several extending resources, which will then have different properties like the radius of the circle or extents of the square, and then the user can select which resource they want". However, the implications of the resource being stored separately and not duplicating in a linked way was likely not considered at all. And I suspect a similar thought-process happened with a number of other nodes, and keeps happening as new nodes get developed.

*Essentially, the Resource functionality is being abused for giving a Node different modes of operation!*

Now, there are nodes that have **modes of operation** functionality implemented, and conditionally show some properties. Examples include the Light nodes that show properties of shadows only if shadows are enabled (though this is a boolean condition), or the Rotation Mode and Rotation order of a Transform (which allows manipulating one and the same property but with different input fields). However, the implementation of that is always quirky, and hard to find and understand. Hence, the pattern of just using Resources, is much more common. 

A different approach is used by CSGShapes, which is made abstract and split into different extending Nodes alltogether. This however increases the barrier for adding a new shape, as it would further clutter the Node list, and switching between shapes is a big hassle.

It should be understood that this architectural problem not only affects how the built-in nodes are used, but also the custom nodes that users create for their games are encouraged to abuse the resource functionality in this way. 
Apart from the duplication dilemma, the resource workaround is also problematic because resources are frequently nested inside other resources that yet again reside inside a resource. Apart from cluttering the UI, it also makes one lose track of which property belongs to which resource and where each one of them is actually stored. 
I am aware that conditionally exposed properties are possible with toolscripts by manipulating the property-list, but applying this is always error-prone, and requires many lines of code that clutter the actual functionality, and so I avoid it whenever I can. 

I would thus propose to think of the following problem:
**How can we develop a *mode of operation* selector for a node, that conditionally exposes properties that are, however, stored with the same node?** Applying this feature should be conceptually and implementationally easy, it shall not clutter code, and should be available equally for built-in and custom nodes/scripts.

**Comment by daniel080400** [comment]
I once tried to solve this issue with PR godotengine/godot#107570, but the discussion brings up several concerns about confusing users. After reading @MajorMcDoom 's comment, I get the idea that maybe we should think outside of a Resource. Like primitive collision shapes should not be a resource, but custom collision (convex, concave) shapes should.

As I assume 90% of the complain is focusing on primitive collision shapes now. Can we implement some derived class of Area3D and StaticBody3D, some common shape like **Area3DBox**, **Area3DSphere**, which does not use a shape resource, but instead has it shape definition parameterized right at the Node level?

**Comment by MajorMcDoom** [comment]
> I once tried to solve this issue with PR [godotengine/godot#107570](https://github.com/godotengine/godot/pull/107570), but the discussion brings up several concerns about confusing users. After reading [@MajorMcDoom](https://github.com/MajorMcDoom) 's comment, I get the idea that maybe we should think outside of a Resource. Like primitive collision shapes should not be a resource, but custom collision (convex, concave) shapes should.
> 
> As I assume 90% of the complain is focusing on primitive collision shapes now. Can we implement some derived class of Area3D and StaticBody3D, some common shape like **Area3DBox**, **Area3DSphere**, which does not use a shape resource, but instead has it shape definition parameterized right at the Node level?

It's not just collisions. I have the same complaint about `Path3D` + `Curve3D`, `MultiMeshInstance3D/2D` + `MultiMesh`.

As for the solution, it's a huge rabbit hole, because even further underneath, at the Server level of Godot, everything uses resources and RIDs, so it would be a huge refactor.

And I'm sure even despite all this, there will still be some people who feel like resources in these contexts offer enough power to be worth the inconveniences. 😅

**Comment by Parsnip** [comment]
> Hi all - Someone posted something about this on Reddit and I think the solution could be fairly simple:
> 
>     * In the editor, In the context menu whilst clicking on a node, replace Duplicate with "Duplicate and link" and "Duplicate and make unique"
> 


This feels like the obvious, most straight forward solution to a seemingly common problem/point of confusion (at least from what I've seen when I was looking this up earlier myself). It's already possible to make resources unique afterwards, it doesn't seem like it needs to be more complicated than this.
Unless there's a limit on how long the text can be in the context menu, you could even be more explicit on what the option does. Like 'Duplicate and link resources' and 'Duplicate and make resources unique' or something.

**Comment by MajorMcDoom** [comment]
> > Hi all - Someone posted something about this on Reddit and I think the solution could be fairly simple:
> > ```
> > * In the editor, In the context menu whilst clicking on a node, replace Duplicate with "Duplicate and link" and "Duplicate and make unique"
> > ```
> 
> This feels like the obvious, most straight forward solution to a seemingly common problem/point of confusion (at least from what I've seen when I was looking this up earlier myself). It's already possible to make resources unique afterwards, it doesn't seem like it needs to be more complicated than this. Unless there's a limit on how long the text can be in the context menu, you could even be more explicit on what the option does. Like 'Duplicate and link resources' and 'Duplicate and make resources unique' or something.

This wouldn't be ideal at all for anything beyond a single node (like a hierarchy), or a node with many properties. With those menu options, either all resources are duplicated or all resources are linked, which is quite arbitrary.

At least with the current behaviour, things can be made into unique resources on a selective basis. If you make everything into duplicates, you can't just selectively relink some of them again.

It's a very destructive process, which would incur invisible performance-killing costs from duplication because beginners would just keep using it liberally without understanding the full ramifications simply because it results in, say, their collision shape not getting synced.

**Comment by SanNull** [comment]
Since it's preferable to make Resources unique on a case-by-case basis, wouldn't it be better to extend `local_to_scene`  functionality so it also accounts for duplicated nodes?

This ensure users can comfortably duplicate needed resources without either linking everything or shooting themselves in the foot by make multiple instances of the same thing.

**Comment by Parsnip** [comment]
> > > Hi all - Someone posted something about this on Reddit and I think the solution could be fairly simple:
> > > ```
> > > * In the editor, In the context menu whilst clicking on a node, replace Duplicate with "Duplicate and link" and "Duplicate and make unique"
> > > ```
> > 
> > 
> > This feels like the obvious, most straight forward solution to a seemingly common problem/point of confusion (at least from what I've seen when I was looking this up earlier myself). It's already possible to make resources unique afterwards, it doesn't seem like it needs to be more complicated than this. Unless there's a limit on how long the text can be in the context menu, you could even be more explicit on what the option does. Like 'Duplicate and link resources' and 'Duplicate and make resources unique' or something.
> 
> This wouldn't be ideal at all for anything beyond a single node (like a hierarchy), or a node with many properties. With those menu options, either all resources are duplicated or all resources are linked, which is quite arbitrary.
> 
> At least with the current behaviour, things can be made into unique resources on a selective basis. If you make everything into duplicates, you can't just selectively relink some of them again.
> 
> It's a very destructive process, which would incur invisible performance-killing costs from duplication because beginners would just keep using it liberally without understanding the full ramifications simply because it results in, say, their collision shape not getting synced.

I would argue that all resources being either linked or not is the actual opposite of arbitrary. It's predictable behavior and it says what it does on the tin. I'd say that for beginners things being more approachable and predictable is much more important than having to maybe worry about potentially losing performance 5 years down the line if they end up shipping a game. Behavior that _feels_ unpredictable (regardless of whether it actually is) is hugely demotivating.

**Comment by MajorMcDoom** [comment]
> I would argue that all resources being either linked or not is the actual opposite of arbitrary. It's predictable behavior and it says what it does on the tin. I'd say that for beginners things being more approachable and predictable is much more important than having to maybe worry about potentially losing performance 5 years down the line if they end up shipping a game. Behavior that _feels_ unpredictable (regardless of whether it actually is) is hugely demotivating.

Teaching bad practices and obscuring concepts isn't a win. It's not just performance, it's bugs too. If you read the previous comments, you will see that there are tons of resources that should be copied by reference, not duplicated. We don't want to trade a whole bunch of reports about this issue for a whole bunch of reports about the opposite issue: "Help, after duplicating, whenever I change my material, the cloned objects don't change!?"

I'm not arguing for not having a solution at all, but allowing a shortcut to blindly and recursively duplicate every nested resource on a complex hierarchy is haphazard. Like setting your whole house on fire to kill a fly. We can't take for granted just how much stuff is expected to be linked.

A good solution should clarify, instead of obscuring the fact that resources are always copied by reference, with uniqueness being an additional, optional step. We need something that makes that additional step easier, or more context-aware. An all-or-nothing UX solution would just exacerbate misunderstanding.

**Comment by Parsnip** [comment]
For most people the concept of "making a copy" means that the original and the copy are unique and that changing either of them afterwards will not affect the other one. So I really don't see even a slightest possibility of an influx of "Help, after duplicating, whenever I change my material, the cloned objects don't change!?" style issues. If the user is completely ignoring what the options say they do, well that's just a different problem altogether isn't it.

For actual beginners, they are not going to have complex hierarchies to duplicate, or hundreds of cloned objects with unique materials or have projects where "misuse" of the unique copy option would actually matter in a larger sense, it's a literal non-issue.

For non-beginners, at some point they would learn when it would be useful for them to use unique copy and and when not to use it.
Simply "allowing" a shortcut for it doesn't automatically mean that that would be the only option people would use. At some point the user should be trusted to make their own (good or bad) choices with the software and if they are making unique copies of things they probably shouldn't be just because the shortcut was there, well, that's again kind of a different issue.


But to be completely honest it's all a bit moot isn't it.
Considering this GH issue is from 2019 and there's no real traction on it one direction (true instance) or another (true unique copy) after all these years, sadly I really don't expect anything to change in this particular matter.

**Comment by MajorMcDoom** [comment]
> For most people the concept of "making a copy" means that the original and the copy are unique and that changing either of them afterwards will not affect the other one. So I really don't see even a slightest possibility of an influx of "Help, after duplicating, whenever I change my material, the cloned objects don't change!?" style issues. If the user is completely ignoring what the options say they do, well that's just a different problem altogether isn't it.

Making a copy of an object is not the same as making a copy of every resource it uses. If your argument is the expected uniqueness of collision shapes, I absolutely agree with you on that, but we're talking everything from scripts to materials to sound randomizers to curves to textures to virtually everything that isn't a node. One cannot just generalize the interpretation of the word "copy" to cover the duplication of things that have always been (including in other engines) copied by reference. You can't just go off of "what would a complete non-game-developer think?" for everything.
 
> For actual beginners, they are not going to have complex hierarchies to duplicate, or hundreds of cloned objects with unique materials or have projects where "misuse" of the unique copy option would actually matter in a larger sense, it's a literal non-issue.

I think you're vastly underestimating the complexity of "beginner" games. It does not take hundreds of nodes to present a problem, it just takes a couple. I also gave an example of something that is easily found in beginner projects (copying an object resulting in material breakages). If it wasn't clear enough, imagine a beginner cloned, I don't know, a goblin, and then decides they want their goblins blue instead of green. They go and change their `goblin.material`, but only the original goblin changes color. This isn't advanced. And at no point did they unreasonably misinterpret what "copy" means.

> But to be completely honest it's all a bit moot isn't it. Considering this GH issue is from 2019 and there's no real traction on it one direction (true instance) or another (true unique copy) after all these years, sadly I really don't expect anything to change in this particular matter.

Pace is slow on a lot of things yeah, but I don't think that's surprising, especially for controversial things. I do however think that a new path forward is to recognize that there are specific use cases that require special care. I believe the unifying factor is "spatial locality". That is, there are many things that Godot uses resources for which are typically associated with a specific locational presence. Collision shapes and paths are both examples of this. When you duplicate these, 99% of the time, they *should* be duplicated anew. A good solution would be to make the nodes that house these resources by default force uniqueness.

**Comment by Parsnip** [comment]
> Making a copy of an object is not the same as making a copy of every resource it uses. If your argument is the expected uniqueness of collision shapes, I absolutely agree with you on that, but we're talking everything from scripts to materials to sound randomizers to curves to textures to virtually everything that isn't a node. One cannot just generalize that the interpretation of the word "copy" to cover the duplication of things that have always been (including in other engines) copied by reference. You can't just go off of "what would a complete non-game-developer think?" for everything.

I think I'll just have to disagree. From what I've seen the people who have the biggest issue with the way things work right now are beginners, non-game-developers. It doesn't matter to them that making a copy of an object is not the same as making a copy of every resource it uses.

> I think you're vastly underestimating the complexity of "beginner" games. It does not take hundreds of nodes to present a problem, it just takes a couple. I also gave an example of something that is easily found in beginner projects (copying an object resulting in material breakages). If it wasn't clear enough, imagine a beginner cloned, I don't know, a goblin, and then decides they want their goblins blue instead of green. They go and change their `goblin.material`, but only the original goblin changes color. This isn't advanced. And at no point did they unreasonably misinterpret what "copy" means.

Again hard disagree. We seem to have a very different perception of what a beginner is.
Matching the expectation of what copying or duplicating means to most people is the whole argument. Goblin.material not changing the color of goblin.2 or goblin.4 is expected, that's what goblin.material.2 and goblin.material.4 are for. They created a unique copy and as expected each copy has a wholly unique material.
Obviously it wouldn't be ideal in the long term to always create new unique materials (for example) but I genuinely don't think that's an actual real issue.

It feels like a theoretical "if given the option the user is only ever going to use the unique copy option forever, never learns anything else about how the engine, materials, instancing, literally how anything works, never iterates on the first objects they worked on."

Like yes, I get that if given the option it could potentially encourage bad practices, but I really don't think that's what this would actually do. In my experience, when given multiple clearly labeled options, it's more likely that the user ends up actually looking into what the options do. "Oh there's more than one way to duplicate things, maybe I should look into that." It's a nudge to learn more. In this case it would help if the context menu had tooltips with direct links to docs though.

But instead we now have "Oh I know what duplicate means. Wait why is this not behaving the way I expected." It creates unnecessary frustration that _might_ make them seek answers, but I've personally found that gently nudging people to discover on their own works better.
